File: 90. Subsets II.py
class Solution:
    def subsetsWithDup(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        if len(nums) == 0:
            return [[]]
        if len(nums) == 1:
            return [nums, []]
        l1 = self.subsetsWithDup(nums[1:])
        le = len(l1)
        for i in range(le):
            l3 = sorted([nums[0]] + l1[i])
            if l3 not in l1:
                l1.append(l3)
        return l1


sol = Solution()
print(sol.subsetsWithDup([1, 2, 2, 3]))
print(sol.subsetsWithDup([1, 2, 3]))
print(sol.subsetsWithDup([1, 2]))

File: 90. Subsets II.cpp
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> ans{{}};
        size_t begin = 0;
        for (size_t i=0; i < nums.size(); ++i) {
            if (i == 0 || nums[i] != nums[i-1]) begin = 0;
            for (size_t end = ans.size(); begin < end; ++begin) {
                vector<int> tmp(ans[begin]);
                tmp.push_back(nums[i]);
                ans.push_back(tmp);
            }
        }
        return ans;
    }
};

int main() {
    Solution a;
    vector<int> v{1,2,2};
    vector<vector<int>> ans = a.subsetsWithDup(v);
    cout << "Size: " << ans.size() << endl;
    for (auto &vec: ans) {
        for (auto element: vec) {
            cout << element << ' ';
        }
        cout << endl;
    }
    return 0;
}
